File: 416.PartitionEqualSubsetSum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        target = sum(nums)
        if target % 2 != 0:
            return False
        else:
            target = target // 2

        sums = set()
        sums.add(0)
        
        for i in range(len(nums)):
            nextDP = set()
            for s in sums:
                if nums[i] + s == target:
                    return True
                nextDP.add(s)
                nextDP.add(nums[i] + s)
            sums = nextDP
        return False
            

File: 416.PartitionEqualSubsetSum.go
func canPartition(nums []int) bool {
    sum := sum(nums)
    if sum % 2 != 0{
        return false
    }
    target := sum / 2
    set := make(map[int]bool)
    set[0] = true
    for _,i := range nums{
        nextSet := make(map[int]bool)
        for s := range set{
            if (s + i) == target{
                return true
            }
            nextSet[s] = true
            nextSet[s + i] = true
        }
        set = nextSet
    }

    return false
}

func sum(nums []int) int{
    res := 0
    for _, i := range nums{
        res += i
    }
    return res
}
